### PR TITLE
RecoLocalTracker/SiPixelRecHits: changes to vectorize loops in VVIObjF::VVIObjF constructor

### PR DESCRIPTION
**Pull request by gartung**
Using Intel Advisor to profile workflow r39634.21 with a CMSSW_12_5_LTO integration build release showed the top time consuming loop with floating point operation is in VVIObjF::VVIObjF constructor. 
```
"1381";"[loop in VVIObjF::VVIObjF at log.h:179]";"41.230s";"28.288s";"Scalar";"";"";"";"";"log.h:179";"libRecoLocalTrackerSiPixelRecHits.so"
```
With this change the loop calculations are vectorized and the time is reduced.
```
"1772";"[loop in VVIObjF::VVIObjF at log.h:179]";"3.824s";"3.824s";"Vectorized (Body)";"";"SSE2";"";"";"log.h:179";"libRecoLocalTrackerSiPixelRecHits.so"
```

Further profiling is being done with CMSSW_12_5_X integration build release.

### CODE REVIEW

File: RecoLocalTracker/SiPixelRecHits/src/VVIObjF.cc [hunk 1]
```diff
@@ -105,33 +105,81 @@ VVIObjF::VVIObjF(float kappa, float beta2, int mode) : mode_(mode) {
   if (mode_ == 0) {
     a_[n - 1] = omega_ * .31830988654751274f;
   }
-  q = -1.;
-  q2 = 2.;
+
+  float q_[n];
```
Comment (slava77) [inline RecoLocalTracker/SiPixelRecHits/src/VVIObjF.cc:109]: it would be nice to not mix the trailing underscore for locals and data members.

Comment (gartung) [reply RecoLocalTracker/SiPixelRecHits/src/VVIObjF.cc:109]: What would you suggest? I thought  of `_array` to distinguish them from the local variables they are replacing.

Comment (slava77) [reply RecoLocalTracker/SiPixelRecHits/src/VVIObjF.cc:109]: leading underscore is against the naming rules. would a trailing cap (e.g. V or A to denote auto/vectorization) work.

**Comment by gartung** [comment]
enable profiling

**Comment by gartung** [comment]
please test

**Comment by tvami** [comment]
(as one of the maintainers of that object) Can we please have some high stat samples where the pixel residuals and the charge distributions are checked? I dont think the results in these low stat RelVals in Jenkins are going to be satisfactory to ensure nothing is changed. Please note Vavilov describes the tail of the charge distribution much better, so that's the part I'd like to see no changes in the comparisons.

**Comment by gartung** [comment]
The ROOT implementation of the fast Vavilov distribution seems to be structured to be vectorized.
https://root.cern/doc/master/VavilovFast_8cxx_source.html

**Comment by gartung** [comment]
Closing until I can prove that the vectorized calculation is even close to the iterative calculation.

**Comment by tvami** [comment]
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9899c9/27126/runTheMatrix-results/11834.0_TTbar_14TeV+2021PU+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoNanoPU+HARVESTNanoPU/step3_TTbar_14TeV+2021PU+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoNanoPU+HARVESTNanoPU.log

```
ProbQ outside the bounds of the quality word: ProbQ = 1+-1.10252 and QualityWord = 3910 --> Now defaulting to ProbQ = 0.0
```
this should not happen for sure

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9899c9/27126/runTheMatrix-results/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA/step3_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA.log

similarly 
```
%MSG-e SiPixelTemplateReco:  TrackProducer:detachedQuadStepTracks  26-Aug-2022 19:31:50 CEST Run: 1 Event: 3
illegal chi2xmin normalization (1) = -0.687319 
%MSG
```

**Comment by gartung** [comment]
Output comparing distribution from vectorized loop `aV[], bV[]` vs iterative loop `a_[], b_[]`
```
aV[0] -4.26422e-07a_[0] -4.26422e-07
aV[1] 5.58271e-07 a_[1] 5.58271e-07
aV[2] -7.29633e-07 a_[2] -7.29633e-07
aV[3] 9.51455e-07 a_[3] 9.51455e-07
aV[4] -1.23712e-06 a_[4] -1.23712e-06
aV[5] 1.60263e-06 a_[5] 1.60263e-06
aV[6] -2.06654e-06 a_[6] -2.06654e-06
aV[7] 2.64928e-06 a_[7] 2.64928e-06
aV[8] -3.37167e-06 a_[8] -3.37167e-06
aV[9] 4.25185e-06 a_[9] 4.25185e-06
aV[10] -5.29948e-06 a_[10] -5.29948e-06
aV[11] 6.50627e-06 a_[11] 6.50627e-06
aV[12] -7.83e-06 a_[12] -7.83e-06
aV[13] 9.16886e-06 a_[13] 9.16886e-06
aV[14] -1.03228e-05 a_[14] -1.03228e-05
aV[15] 1.09331e-05 a_[15] 1.09331e-05
aV[16] -1.03939e-05 a_[16] -1.03939e-05
aV[17] 7.72569e-06 a_[17] 7.72569e-06
aV[18] -1.39197e-06 a_[18] -1.39197e-06
aV[19] -1.09472e-05 a_[19] -1.09472e-05
aV[20] 3.27566e-05 a_[20] 3.27566e-05
aV[21] -6.90421e-05 a_[21] -6.90421e-05
aV[22] 0.00012681 a_[22] 0.00012681
aV[23] -0.000215511 a_[23] -0.000215511
aV[24] 0.000347348 a_[24] 0.000347348
aV[25] -0.000537008 a_[25] -0.000537008
aV[26] 0.000800216 a_[26] 0.000800216
aV[27] -0.0011501 a_[27] -0.0011501
aV[28] 0.00158896 a_[28] 0.00158896
aV[29] -0.00209239 a_[29] -0.00209239
aV[30] 0.00258076 a_[30] 0.00258076
aV[31] -0.00286858 a_[31] -0.00286858
aV[32] 0.00257985 a_[32] 0.00257985
aV[33] -0.0010154 a_[33] -0.0010154
aV[34] -0.00305453 a_[34] -0.00305453
aV[35] 0.0116963 a_[35] 0.0116963
aV[36] -0.0282074 a_[36] -0.0282074
aV[37] 0.0575693 a_[37] 0.0575693
aV[38] -0.106852 a_[38] -0.106852
aV[39] 0.184624 a_[39] 0.184624
aV[40] 0.75215 a_[40] 0.75215
```

```
bV[0] 7.83574e-08b_[0] 7.83574e-08
bV[1] -1.22857e-07 b_[1] -1.22857e-07
bV[2] 1.90679e-07 b_[2] 1.90679e-07
bV[3] -2.9312e-07 b_[3] -2.9312e-07
bV[4] 4.46563e-07 b_[4] 4.46563e-07
bV[5] -6.74646e-07 b_[5] -6.74646e-07
bV[6] 1.01118e-06 b_[6] 1.01118e-06
bV[7] -1.50422e-06 b_[7] -1.50422e-06
bV[8] 2.22157e-06 b_[8] 2.22157e-06
bV[9] -3.25798e-06 b_[9] -3.25798e-06
bV[10] 4.74469e-06 b_[10] 4.74469e-06
bV[11] -6.86164e-06 b_[11] -6.86164e-06
bV[12] 9.85203e-06 b_[12] 9.85203e-06
bV[13] -1.40402e-05 b_[13] -1.40402e-05
bV[14] 1.98505e-05 b_[14] 1.98505e-05
bV[15] -2.78243e-05 b_[15] -2.78243e-05
bV[16] 3.86322e-05 b_[16] 3.86322e-05
bV[17] -5.30668e-05 b_[17] -5.30668e-05
bV[18] 7.19994e-05 b_[18] 7.19994e-05
bV[19] -9.62765e-05 b_[19] -9.62765e-05
bV[20] 0.000126492 b_[20] 0.000126492
bV[21] -0.000162566 b_[21] -0.000162566
bV[22] 0.000203032 b_[22] 0.000203032
bV[23] -0.000243805 b_[23] -0.000243805
bV[24] 0.000276244 b_[24] 0.000276244
bV[25] -0.00028422 b_[25] -0.00028422
bV[26] 0.000239673 b_[26] 0.000239673
bV[27] -9.65036e-05 b_[27] -9.65036e-05
bV[28] -0.000217272 b_[28] -0.000217272
bV[29] 0.000808951 b_[29] 0.000808951
bV[30] -0.00182916 b_[30] -0.00182916
bV[31] 0.00347179 b_[31] 0.00347179
bV[32] -0.00595329 b_[32] -0.00595329
bV[33] 0.0094367 b_[33] 0.0094367
bV[34] -0.0138262 b_[34] -0.0138262
bV[35] 0.0182811 b_[35] 0.0182811
bV[36] -0.0200421 b_[36] -0.0200421
bV[37] 0.0112811 b_[37] 0.0112811
bV[38] 0.0314543 b_[38] 0.0314543
bV[39] -0.209585 b_[39] -0.209585
bV[40] 1.09514e-26 b_[40] 0
```

**Comment by gartung** [comment]
please test

**Comment by gartung** [comment]
VVIObjF constructor moves from #91 to #110 in time
https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_5_X_2022-08-26-2300/el8_amd64_gcc10/profiling/39634.21/igprofCPU_step3/91
https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_5_X_2022-08-25-2300/el8_amd64_gcc10/profiling/39634.21/PR-9899c9/27137/igprofCPU_step3/110

**Comment by gartung** [comment]
enable high-stats

**Comment by gartung** [comment]
enable high_stats

**Comment by gartung** [comment]
please test

**Comment by slava77** [comment]
in the comparison results in https://github.com/cms-sw/cmssw/pull/39218#issuecomment-1228968302 when the code was broken, there are no differences in monitored quantities.

@tvami @cms-sw/trk-dpg-l2 @cms-sw/tracking-pog-l2 (partly a q to self)
where is the probQ used?
It seems like it is not used during tracking.
Can we disable this computation in the regular CPE calls, and postpone the computation to the point where it's actually needed?
Even after this PR, it's still 2% of reco time that seems to mainly go to /dev/null

**Comment by mmusich** [comment]
It goes into `clusterProbability`:

> where is the probQ used?

https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/DataFormats/TrackerRecHit2D/src/SiPixelRecHit.cc#L13

which is used in DQM, Alignment and a smattering of related applications.
I am not sure if there's any EXO related analysis that uses it as well directly, @tvami might know better.

**Comment by tvami** [comment]
> I am not sure if there's any EXO related analysis that uses it as well directly

Yes, EXO HSCP uses it directly as a discriminator. It could be good for other stuff too but that was not tested (yet)

**Comment by mmusich** [comment]
> Can we disable this computation in the regular CPE calls, and postpone the computation to the point where it's actually needed?

To be clear,  for the applications I am aware of, it would make sense to calculate it only at track refit time and not directly during the reconstruction.

**Comment by gartung** [comment]
enable profiling

**Comment by mandrenguyen** [comment]
+reconstruction
Technical.  Diffs in reco comparisons appears negligible

**Comment by tvami** [comment]
assign trk-dpg

**Comment by mmusich** [comment]
> you have been requested to review this Pull request/Issue and eventually sign

the last round of tests (also the logs) look [totally clean](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-27-1100+9899c9/52418/11834.0_TTbar_14TeV+2021PU+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoNanoPU+HARVESTNanoPU/PixelPhase1.html) and we have in principle monitoring of the cluster probability in the offline DQM.
Not sure if some specific validation would still be required here, but I would be in favor of just signing.

**Comment by gartung** [comment]
This is the branch I used for the comparison of the arrays calculated with the iterative loop and with the vectorized loop. It produces a lot of output so it should only be used on one event.
https://github.com/cms-sw/cmssw/compare/master...gartung:cmssw:gartung-vviobjf-vectorize-test

**Comment by mmusich** [comment]
+1
  * based on https://github.com/cms-sw/cmssw/pull/39218#issuecomment-1231563314

**Comment by perrotta** [comment]
+1
- I understand that some further optimization is still possible, based on https://github.com/cms-sw/cmssw/pull/39218#issuecomment-1229260140 and https://github.com/cms-sw/cmssw/pull/39218#issuecomment-1229266730

**Pull request merged**